### DISCUSSION TITLE
How to setup ZLS correctly?

### DISCUSSION THREAD

**Discussion by ghost** [zigtools/zls#1715] [Q&A]
So I want to try out Zig and already got the compiler working on my Windows 11. I programm over VS Code and installed all the extensions needed and also wanted to Debug my first programs. While Zig just has test and not a real debbuger it comes with, I found a guide on Reddit to set up a debbuging option for Zig. 

Link: https://www.reddit.com/r/Zig/comments/hqjw0n/how_i_setup_zig_on_windows/

I also installed zls in C:\Program\zig\zls\zig-cache\bin and put the zls.exe and zls.json file in there. But in cmd the zls command always throws me an error that it cant find the .json file. 

So since I'm bloody new to Zig any help will be apreciated :)

**Comment by Vexu** [comment]
If you install the [offical VSCode Zig extension](https://marketplace.visualstudio.com/items?itemName=ziglang.vscode-zig) you can have the extension manage installing both Zig and ZLS for you.

**Reply by ghost** [reply]
This sounds like an great solution. Could U explain how to do so? I'm not quite sure where to find the installing Option?

**Reply by techatrix** [reply]
When you first install the extension, you should get prompted to install Zig and ZLS.
You can manually do this by opening the Command Palette in VS Code and run the "Zig Setup: Install Zig" command. After selecting the Zig version, you should also be asked if you would like to install ZLS. If not, you can use the "Zig Language Server: Install Server" command.

**Reply by ghost** [reply]
Ok I've installed Zig over the extension, but I now I get the error that Zig isnt actually installed. I installed a second time and still get the error that it isnt installed as well zls. Is there another configuration I have to do before using Zig and zls?

**Reply by techatrix** [reply]
This shouldn't happen...
Could you open your `settings.json` with the `Preferences: Open User Settings (JSON)` command and remove any settings that mentions Zig or ZLS. Then restart VS Code. You should then be asked if you would like to install Zig again once you opened a `.zig` file.

This is the popup you should see:
![Screenshot from 2024-01-15 21-37-26](https://github.com/zigtools/zls/assets/19954306/3bbb7541-7583-4068-b909-55e766c067ff)

**Reply by ghost** [reply]
Ok I this worked fine. I got the popup window which you also have

**Reply by ghost** [reply]
When I try to install it again I'll get this error: Unable to install Zig: Error: ENOTEMPTY: directory not empty, rmdir '\\?\c:\Users\me\AppData\Roaming\Code\User\globalStorage\ziglang.vscode-zig\zig_install'

**Reply by techatrix** [reply]
There may be an old installation of zig in the global vscode storage that is being removed for some reason.
I would suggest to remove the `c:\Users\nick\AppData\Roaming\Code\User\globalStorage\ziglang.vscode-zig\` directory and try to install again.

---

@Vexu any Idea how this is happening?
I suspect it is coming from here:
https://github.com/ziglang/vscode-zig/blob/bf09c1f3cc2e4bcc385340fe2fe50ac2ab63e3d4/src/zigSetup.ts#L58C15-L58C15
But it is recursive and force so why is it failing with directory not empty?

Edit:
Should that line call `fs.rmdirSync` instead of `fs.rmSync` since the path is a directory and not a file

**Reply by ghost** [reply]
So I deleted everything in that folder, was asked again to install Zig and installed it again with the same results. Maybe a config file is false initilised? I get the same errors after installing zig. When trying to compile the zig file in VS, it gives me an error that it cant find the command zig.

**Reply by techatrix** [reply]
Zig is not available in your terminal unless you add the installed Zig executable to your PATH.
The path to the installed Zig executable can be found in your `settings.json`.

**Reply by ghost** [reply]
Ok. Set the Path Variable to my zig and zls installation and zig works now in cmd. When trying to compile the programm I get this error.

error: the following command failed with 'Unexpected':
C:\Users\me\AppData\Local\zig\o\64cfcde33b7f1e851932420ad1ca34d9\main.exe

This is what "zig run" ouputs. 

Edit: The Path to my zig installation is not written in my settings.json
